src/RoomDirectoryModel.cpp

```diff
@@ -0,0 +1,74 @@
+// SPDX-FileCopyrightText: 2021 Nheko Contributors
+//
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+#include "RoomDirectoryModel.h"
+#include "CompletionModelRoles.h"
+
+RoomDirectoryModel::RoomDirectoryModel(std::string &s, QObject *parent)
+                    : QAbstractListModel(parent)
+                    , server_(s)
+{
+    // ??? don't need to do anything special here ???
+}
+
+QHash<int, QByteArray> 
+RoomDirectoryModel::roleNames() const
+{
+    // ??? TODO once I understand what the roles actually mean
+    return {};
+}
+
+QVariant
+RoomDirectoryModel::data(const QModelIndex &index, int role) const
+{
+    if (hasIndex(index.row(), index.column(), index.parent())) {
+        switch (role)
+        {
+        // TODO once I figure out what roles are involved
+
+        // case /* constant-expression */:
+        //     /* code */
+        //     break;
+        
+        // default:
+        //     break;
+        }
+    } else {
+        return {};
+    }
+}
+
+void
+RoomDirectoryModel::fetchMore(const QModelIndex &parent) {
+    // call the GET /_matrix/client/r0/publicRooms endpoint
+    // from mtxclient which returns 
+    http::client()->get_public_rooms(
+                            // Callback<mtx::responses::PublicRooms> cb,
+                            [this](
+                                const mtx::responses::PublicRooms &res,
+                                RequestErr err)
+                                {
+                                    if (err) {
+                                        nhlog::net()->error("Failed to GET public rooms batch - {} - {} - {}",
+                                        mtx::errors::to_string(err->matrix_error.errcode),
+                                        err->matrix_error.error,
+                                        err->parse_error);
+                                    } else {
+                                        // ??? should we keep only the latest chunk, or should out vector
+                                        // of rooms store all rooms from the very first chunk batch???  
+                                        publicRoomsData_ = res.chunk; // ???copying seems like the right thing to do over moving???
+                                        // ??? do we split the chunk fields and populate the displayRoomNames_ etc member fields accordingly ???
```

Comment:
Well, it depends how you want to store them, but make sure to call begin/endInsertRows accordingly: https://doc.qt.io/qt-5/qabstractitemmodel.html#beginInsertRows